File: internal/attachment/filetypes.go
// Copyright (c) 2025 Reliant Labs
package attachment

import (
	"path/filepath"
	"strings"
)

// AttachmentType represents how an attachment should be handled
type AttachmentType string

const (
	// TypeImage - binary image files sent as image blocks to LLM
	TypeImage AttachmentType = "image"
	// TypeFileReference - text files stored as path, content read at send time
	TypeFileReference AttachmentType = "file_reference"
	// TypeDocument - binary document files (PDF, etc.) sent natively to LLMs
	TypeDocument AttachmentType = "document"
	// TypeUnsupported - files that cannot be attached
	TypeUnsupported AttachmentType = "unsupported"
)

// Image extensions supported by Claude API (base64 image blocks)
var ImageExtensions = map[string]bool{
	".jpg":  true,
	".jpeg": true,
	".png":  true,
	".gif":  true,
	".webp": true,
}

// Text file extensions that can be read and sent as text content
var TextExtensions = map[string]bool{
	// Markdown and documentation
	".md":       true,
	".markdown": true,
	".mdx":      true,
	".txt":      true,
	".text":     true,
	".rst":      true,

	// Data formats
	".json":       true,
	".yaml":       true,
	".yml":        true,
	".toml":       true,
	".xml":        true,
	".csv":        true,
	".tsv":        true,
	".ini":        true,
	".conf":       true,
	".cfg":        true,
	".env":        true,
	".properties": true,
	".pdf":        true,
	".docx":       true,

	// Programming languages
	".go":    true,
	".py":    true,
	".js":    true,
	".jsx":   true,
	".ts":    true,
	".tsx":   true,
	".rs":    true,
	".java":  true,
	".kt":    true,
	".scala": true,
	".c":     true,
	".cpp":   true,
	".cc":    true,
	".cxx":   true,
	".h":     true,
	".hpp":   true,
	".hxx":   true,
	".cs":    true,
	".swift": true,
	".rb":    true,
	".php":   true,
	".pl":    true,
	".pm":    true,
	".lua":   true,
	".r":     true,
	".R":     true,
	".m":     true, // Objective-C or MATLAB
	".mm":    true, // Objective-C++
	".zig":   true,
	".nim":   true,
	".v":     true, // V lang
	".d":     true, // D lang
	".dart":  true,
	".ex":    true, // Elixir
	".exs":   true,
	".erl":   true, // Erlang
	".hrl":   true,
	".clj":   true, // Clojure
	".cljs":  true,
	".cljc":  true,
	".fs":    true, // F#
	".fsx":   true,
	".ml":    true, // OCaml
	".mli":   true,
	".hs":    true, // Haskell
	".lhs":   true,
	".elm":   true,
	".purs":  true, // PureScript
	".jl":    true, // Julia

	// Shell and scripts
	".sh":   true,
	".bash": true,
	".zsh":  true,
	".fish": true,
	".ps1":  true, // PowerShell
	".psm1": true,
	".bat":  true,
	".cmd":  true,

	// Web
	".html":   true,
	".htm":    true,
	".css":    true,
	".scss":   true,
	".sass":   true,
	".less":   true,
	".vue":    true,
	".svelte": true,
	".astro":  true,

	// Config and build
	".dockerfile": true,
	".makefile":   true,
	".cmake":      true,
	".gradle":     true,
	".sbt":        true,
	".cabal":      true,
	".gemspec":    true,
	".podspec":    true,

	// SQL and queries
	".sql":     true,
	".graphql": true,
	".gql":     true,

	// Other text formats
	".log":    true,
	".diff":   true,
	".patch":  true,
	".proto":  true,
	".thrift": true,
	".avsc":   true, // Avro schema
	".tf":     true, // Terraform
	".tfvars": true,
	".hcl":    true, // HashiCorp Configuration Language
}

// Special filenames that are text files regardless of extension
var TextFilenames = map[string]bool{
	"Dockerfile":     true,
	"Makefile":       true,
	"CMakeLists.txt": true,
	"Gemfile":        true,
	"Rakefile":       true,
	"Vagrantfile":    true,
	"Procfile":       true,
	"Brewfile":       true,
	".gitignore":     true,
	".gitattributes": true,
	".dockerignore":  true,
	".editorconfig":  true,
	".prettierrc":    true,
	".eslintrc":      true,
	".babelrc":       true,
	".npmrc":         true,
	".nvmrc":         true,
	".yarnrc":        true,
	"LICENSE":        true,
	"README":         true,
	"CHANGELOG":      true,
	"AUTHORS":        true,
	"CONTRIBUTORS":   true,
	"COPYING":        true,
	"NOTICE":         true,
	"VERSION":        true,
}

// GetAttachmentType determines how a file should be handled based on its name
func GetAttachmentType(filename string) AttachmentType {
	// Check special filenames first
	basename := filepath.Base(filename)
	if TextFilenames[basename] {
		return TypeFileReference
	}

	// Get extension (lowercase for comparison)
	ext := strings.ToLower(filepath.Ext(filename))
	if ext == "" {
		return TypeUnsupported
	}

	// Check if it's an image
	if ImageExtensions[ext] {
		return TypeImage
	}

	// Check if it's a text file
	if TextExtensions[ext] {
		return TypeFileReference
	}

	return TypeUnsupported
}

// GetMimeType returns the MIME type for a file extension
func GetMimeType(ext string) string {
	ext = strings.ToLower(ext)
	switch ext {
	// Images
	case ".jpg", ".jpeg":
		return "image/jpeg"
	case ".png":
		return "image/png"
	case ".gif":
		return "image/gif"
	case ".webp":
		return "image/webp"

	// Text and data
	case ".txt", ".text":
		return "text/plain"
	case ".md", ".markdown", ".mdx":
		return "text/markdown"
	case ".json":
		return "application/json"
	case ".xml":
		return "application/xml"
	case ".yaml", ".yml":
		return "application/x-yaml"
	case ".csv":
		return "text/csv"
	case ".pdf":
		return "application/pdf"
	case ".docx":
		return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
	case ".html", ".htm":
		return "text/html"
	case ".css":
		return "text/css"
	case ".js", ".jsx":
		return "application/javascript"
	case ".ts", ".tsx":
		return "application/typescript"

	default:
		return "application/octet-stream"
	}
}

// IsImageType checks if the attachment type is an image
func (t AttachmentType) IsImage() bool {
	return t == TypeImage
}

// IsFileReference checks if the attachment type is a file reference
func (t AttachmentType) IsFileReference() bool {
	return t == TypeFileReference
}

func (t AttachmentType) IsDocument() bool {
	return t == TypeDocument
}

// IsSupported checks if the attachment type is supported
func (t AttachmentType) IsSupported() bool {
	return t != TypeUnsupported
}

// SupportedExtensions returns a human-readable list of supported extensions
func SupportedExtensions() string {
	var exts []string
	for ext := range ImageExtensions {
		exts = append(exts, ext)
	}
	for ext := range TextExtensions {
		exts = append(exts, ext)
	}
	return strings.Join(exts, ", ")
}

File: web/src/lib/filetypes.ts
// Copyright (c) 2025 Reliant Labs

/**
 * File type classification for attachments.
 * Must be kept in sync with internal/attachment/filetypes.go
 */

export type AttachmentType = 'image' | 'document' | 'file_reference' | 'unsupported';

// Image extensions supported by Claude API (base64 image blocks)
export const IMAGE_EXTENSIONS = new Set([
  '.jpg',
  '.jpeg',
  '.png',
  '.gif',
  '.webp',
]);

// Binary document extensions sent to the LLM natively rather than as extracted
// text. PDFs are read on demand (and paginated) via the read_attachment tool.
export const DOCUMENT_EXTENSIONS = new Set([
  '.pdf',
]);

// Text file extensions that can be read and sent as text content
export const TEXT_EXTENSIONS = new Set([
  // Markdown and documentation
  '.md', '.markdown', '.mdx', '.txt', '.text', '.rst',

  // Data formats
  '.json', '.yaml', '.yml', '.toml', '.xml', '.csv', '.tsv',
  '.ini', '.conf', '.cfg', '.env', '.properties', '.docx',

  // Programming languages
  '.go', '.py', '.js', '.jsx', '.ts', '.tsx', '.rs', '.java', '.kt', '.scala',
  '.c', '.cpp', '.cc', '.cxx', '.h', '.hpp', '.hxx', '.cs', '.swift',
  '.rb', '.php', '.pl', '.pm', '.lua', '.r', '.R', '.m', '.mm',
  '.zig', '.nim', '.v', '.d', '.dart', '.ex', '.exs', '.erl', '.hrl',
  '.clj', '.cljs', '.cljc', '.fs', '.fsx', '.ml', '.mli', '.hs', '.lhs',
  '.elm', '.purs', '.jl',

  // Shell and scripts
  '.sh', '.bash', '.zsh', '.fish', '.ps1', '.psm1', '.bat', '.cmd',

  // Web
  '.html', '.htm', '.css', '.scss', '.sass', '.less', '.vue', '.svelte', '.astro',

  // Config and build
  '.dockerfile', '.makefile', '.cmake', '.gradle', '.sbt', '.cabal', '.gemspec', '.podspec',

  // SQL and queries
  '.sql', '.graphql', '.gql',

  // Other text formats
  '.log', '.diff', '.patch', '.proto', '.thrift', '.avsc', '.tf', '.tfvars', '.hcl',
]);

// Special filenames that are text files regardless of extension
export const TEXT_FILENAMES = new Set([
  'Dockerfile', 'Makefile', 'CMakeLists.txt', 'Gemfile', 'Rakefile',
  'Vagrantfile', 'Procfile', 'Brewfile',
  '.gitignore', '.gitattributes', '.dockerignore', '.editorconfig',
  '.prettierrc', '.eslintrc', '.babelrc', '.npmrc', '.nvmrc', '.yarnrc',
  'LICENSE', 'README', 'CHANGELOG', 'AUTHORS', 'CONTRIBUTORS', 'COPYING', 'NOTICE', 'VERSION',
]);

/**
 * Get the file extension from a filename (lowercase)
 */
function getExtension(filename: string): string {
  const lastDot = filename.lastIndexOf('.');
  if (lastDot === -1 || lastDot === filename.length - 1) {
    return '';
  }
  return filename.slice(lastDot).toLowerCase();
}

/**
 * Get the basename from a path
 */
function getBasename(filepath: string): string {
  const lastSlash = Math.max(filepath.lastIndexOf('/'), filepath.lastIndexOf('\\'));
  return lastSlash === -1 ? filepath : filepath.slice(lastSlash + 1);
}

/**
 * Determine how a file should be handled based on its name
 */
export function getAttachmentType(filename: string): AttachmentType {
  const basename = getBasename(filename);
  
  // Check special filenames first
  if (TEXT_FILENAMES.has(basename)) {
    return 'file_reference';
  }

  const ext = getExtension(filename);
  if (!ext) {
    return 'unsupported';
  }

  // Check if it's an image
  if (IMAGE_EXTENSIONS.has(ext)) {
    return 'image';
  }

  // Check if it's a binary document (PDF) sent natively / read on demand
  if (DOCUMENT_EXTENSIONS.has(ext)) {
    return 'document';
  }

  // Check if it's a text file
  if (TEXT_EXTENSIONS.has(ext)) {
    return 'file_reference';
  }

  return 'unsupported';
}

/**
 * Check if a file type is supported for attachment
 */
export function isFileTypeSupported(filename: string): boolean {
  return getAttachmentType(filename) !== 'unsupported';
}

/**
 * Check if a file is an image type
 */
export function isImageFile(filename: string): boolean {
  return getAttachmentType(filename) === 'image';
}

/**
 * Check if a file is a text/file reference type
 */
export function isTextFile(filename: string): boolean {
  return getAttachmentType(filename) === 'file_reference';
}

/**
 * Get a human-readable error message for unsupported file types
 */
export function getUnsupportedFileMessage(filename: string): string {
  const ext = getExtension(filename);
  if (!ext) {
    return `Files without extensions are not supported. Supported: images (jpg, png, gif, webp) and text files (md, txt, json, code files, etc.)`;
  }
  return `File type "${ext}" is not supported. Supported: images (jpg, png, gif, webp) and text files (md, txt, json, code files, etc.)`;
}

/**
 * Get MIME types for the file input accept attribute
 */
export function getAcceptedMimeTypes(): string {
  // Images
  const imageMimes = ['image/jpeg', 'image/png', 'image/gif', 'image/webp'];
  
  // Text - use wildcard for text/* and common extensions
  const textMimes = [
    'text/*',
    'application/json',
    'application/xml',
    'application/x-yaml',
    'application/pdf',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  ];
  
  // Also include common extensions for file picker
  const extensions = [
    ...Array.from(IMAGE_EXTENSIONS),
    ...Array.from(DOCUMENT_EXTENSIONS),
    ...Array.from(TEXT_EXTENSIONS),
  ];

  return [...imageMimes, ...textMimes, ...extensions].join(',');
}